packages/tui/src/layout.rs

```diff
@@ -1,131 +1,83 @@
+                    }
+                } else {
+                    self.node = Some(stretch.new_node(style, &[]).unwrap());
+                }
 
-            nodes.insert(
-                id,
-                TuiNode {
-                    node,
-                    block_style: RinkStyle::default(),
-                    tui_modifier: TuiModifier::default(),
-                    layout: layout.new_node(style, &[]).unwrap(),
-                },
-            );
-        }
-        VNode::Element(el) => {
-            // gather up all the styles from the attribute list
-            let mut modifier = StyleModifer {
-                style: Style::default(),
-                tui_style: RinkStyle::default(),
-                tui_modifier: TuiModifier::default(),
-            };
+                self.style = style;
+            }
+            VNode::Element(el) => {
+                // gather up all the styles from the attribute list
+                let mut style = Style::default();
 
-            // handle text modifier elements
-            if el.namespace.is_none() {
-                match el.tag {
-                    "b" => apply_attributes("font-weight", "bold", &mut modifier),
-                    "strong" => apply_attributes("font-weight", "bold", &mut modifier),
-                    "u" => apply_attributes("text-decoration", "underline", &mut modifier),
-                    "ins" => apply_attributes("text-decoration", "underline", &mut modifier),
-                    "del" => apply_attributes("text-decoration", "line-through", &mut modifier),
-                    "i" => apply_attributes("font-style", "italic", &mut modifier),
-                    "em" => apply_attributes("font-style", "italic", &mut modifier),
-                    "mark" => apply_attributes(
-                        "background-color",
-                        "rgba(241, 231, 64, 50%)",
-                        &mut modifier,
-                    ),
-                    _ => (),
+                for &Attribute { name, value, .. } in el.attributes {
+                    apply_layout_attributes(name, value, &mut style);
                 }
-            }
 
-            for &Attribute { name, value, .. } in el.attributes {
-                apply_attributes(name, value, &mut modifier);
-            }
+                // the root node fills the entire area
+                if el.id.get() == Some(ElementId(0)) {
+                    apply_layout_attributes("width", "100%", &mut style);
+                    apply_layout_attributes("height", "100%", &mut style);
+                }
 
-            // Layout the children
-            for child in el.children {
-                collect_layout(layout, nodes, vdom, child);
-            }
+                // Set all direct nodes as our children
+                let mut child_layout = vec![];
+                for l in children {
+                    child_layout.push(l.node.unwrap());
```

Comment:
this panics on early return from component (`l.node == None`).
is skipping such children correct?

```rust
let child_layout: Vec<Node> = children.filter_map(|l| l.node).collect();
```

Reply:
Node should only be none if update_state hasn't been called. The function is resolved from the lowest node in the tree upward, so it always will be some within this function. I should change the Option to an enum with uninitiated to make the code more readable.

Reply:
It might be worth changing this at the api level so that update takes a mutable reference to a possibly uninitialized self instead of requiring the default trait bounds

Reply:
> so it always will be some within this function

i'm not familiar with this codebase yet, just experimenting. minimal example:

```rust
use dioxus::prelude::*;

fn main() {
    dioxus::tui::launch(app)
}

fn app(cx: Scope) -> Element {
    cx.render(rsx! {
        div {
            component()
        }
    })
}

#[inline_props]
fn component(cx: Scope) -> Element {
    None
}
```

```
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', /home/univerz/.cargo/git/checkouts/dioxus-9d150aea5a02f9fc/3dd18b2/packages/tui/src/layout.rs:64:46
stack backtrace:
   0:     0x55d8526906dd - std::backtrace_rs::backtrace::libunwind::trace::hda39d6860b2d6f70
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x55d8526906dd - std::backtrace_rs::backtrace::trace_unsynchronized::ha1bfdeb508735a0a
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x55d8526906dd - std::sys_common::backtrace::_print_fmt::h4e9e937d66742a36
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x55d8526906dd - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h73a948c4553c4f51
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x55d8526b0a9c - core::fmt::write::ha91e2a0e6cdcb7d1
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/core/src/fmt/mod.rs:1194:17
   5:     0x55d85268ca31 - std::io::Write::write_fmt::heda409ac36864e96
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/io/mod.rs:1655:15
   6:     0x55d852691fe5 - std::sys_common::backtrace::_print::he8c3809f69ea2342
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x55d852691fe5 - std::sys_common::backtrace::print::hecbd7b926628aa09
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x55d852691fe5 - std::panicking::default_hook::{{closure}}::h8fbf7884094b6f2e
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/panicking.rs:295:22
   9:     0x55d852691c59 - std::panicking::default_hook::h6334d4a28312e99b
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/panicking.rs:314:9
  10:     0x55d852692692 - std::panicking::rust_panic_with_hook::h73dbf64f0060aa58
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/panicking.rs:698:17
  11:     0x55d852692539 - std::panicking::begin_panic_handler::{{closure}}::h57f62f8b0c4aa3d9
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/panicking.rs:586:13
  12:     0x55d852690b94 - std::sys_common::backtrace::__rust_end_short_backtrace::h3194a80099ffb120
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/sys_common/backtrace.rs:138:18
  13:     0x55d8526922a9 - rust_begin_unwind
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/panicking.rs:584:5
  14:     0x55d8526afab3 - core::panicking::panic_fmt::h77eb0c9a0a627689
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/core/src/panicking.rs:142:14
  15:     0x55d8526af8fd - core::panicking::panic::h72eee331f65d5eb5
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/core/src/panicking.rs:48:5
  16:     0x55d852601c89 - core::option::Option<T>::unwrap::hda627f64fff24639
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/core/src/option.rs:755:21
  17:     0x55d852601c89 - <dioxus_tui::layout::StretchLayout as dioxus_native_core::real_dom::BubbledUpState>::reduce::h1bc15ed5d5acd0cd
                               at /home/univerz/.cargo/git/checkouts/dioxus-9d150aea5a02f9fc/3dd18b2/packages/tui/src/layout.rs:64:39
  18:     0x55d8525f7b7c - dioxus_native_core::real_dom::RealDom<US,DS>::update_state::h4d5583689fc8dae3
                               at /home/univerz/.cargo/git/checkouts/dioxus-9d150aea5a02f9fc/3dd18b2/packages/native-core/src/real_dom.rs:222:13
  19:     0x55d85260208f - dioxus_tui::launch_cfg::h8208bacb9ec9fc1b
                               at /home/univerz/.cargo/git/checkouts/dioxus-9d150aea5a02f9fc/3dd18b2/packages/tui/src/lib.rs:77:24
  20:     0x55d8525e2133 - core::ops::function::FnOnce::call_once::h77b6b1956a25dc65
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/core/src/ops/function.rs:248:5
  21:     0x55d8525e2133 - std::sys_common::backtrace::__rust_begin_short_backtrace::h65a6c301e7e033ab
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/sys_common/backtrace.rs:122:18
  22:     0x55d8525e2149 - std::rt::lang_start::{{closure}}::h4fe2d00309e56e16
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/rt.rs:145:18
  23:     0x55d852687c0e - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::h8747dac316cfe8b1
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/core/src/ops/function.rs:280:13
  24:     0x55d852687c0e - std::panicking::try::do_call::ha8cfd36c99c35841
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/panicking.rs:492:40
  25:     0x55d852687c0e - std::panicking::try::h44599ca0771bc4e2
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/panicking.rs:456:19
  26:     0x55d852687c0e - std::panic::catch_unwind::hfc289307f73cb8b7
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/panic.rs:137:14
  27:     0x55d852687c0e - std::rt::lang_start_internal::{{closure}}::h9fadec4e4222aa5a
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/rt.rs:128:48
  28:     0x55d852687c0e - std::panicking::try::do_call::h742ca90b47c7614a
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/panicking.rs:492:40
  29:     0x55d852687c0e - std::panicking::try::ha8060cd5fe01b8e1
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/panicking.rs:456:19
  30:     0x55d852687c0e - std::panic::catch_unwind::h3548704cd9c5e356
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/panic.rs:137:14
  31:     0x55d852687c0e - std::rt::lang_start_internal::hc6526c5151dc66ec
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/rt.rs:128:20
  32:     0x55d8525e2298 - main
  33:     0x7ff8d888034a - <unknown>
  34:     0x7ff8d88803fc - __libc_start_main
  35:     0x55d8525e2061 - _start
  36:                0x0 - <unknown>
```

Reply:
Ok, It looks like it is creating a placeholder node for the component, which is not handled in the match statement.

Reply:
This is fixed in the latest version.